### PR TITLE
[release/3.4] Support hyperball optim-method and shardingV1 for adam

### PR DESCRIPTION
**Pull request by xxyux**
### PR Category

Execute Infrastructure

### PR Types

New features

### Description

**What**

Add the **Hyperball** constraint to the `Muon` optimizer so each parameter can be trained in one of four modes, plus a decoupled `muon_epsilon`.

- **`MuonParamInfo.use_hyperball`** (new field, default `False`): combined with `use_muon`, selects the per-parameter mode — **MuonH** (Muon direction + sphere projection), **AdamH** (Adam direction + sphere projection), plain **Muon**,
  plain **AdamW**.
- **`Muon._hyperball_apply`** (new): the shared projection `W_{t+1} = R·Normalize(W_t − η·R·Normalize(u))`, with `R = ‖W‖_F` recomputed every step (per-matrix for 2D, per-expert for 3D). No weight decay; `R` is not stored in the checkpoint.
- **`Muon._adamw_hyperball_update`** (new): the AdamH path — Adam-preconditioned direction fed into `_hyperball_apply`.
- **`muon_epsilon`** (new ctor arg, default `None`): the Newton–Schulz normalization epsilon, decoupled from `adam_epsilon`. When unset it falls back to `adam_epsilon`, so existing behavior is unchanged.
- **`MuonShardingOptimizer`**: classification changed to `keep_whole = use_muon or use_hyperball`. Hyperball params are kept whole on an owner rank (V1) so the global Frobenius norm is computable locally with no extra collective; plain AdamW stays element-wise sharded (V2).

**Why**

Hyperball fixes each weight matrix on a Frobenius-norm sphere (replacing weight decay), which needs the full matrix to compute `‖W‖_F`. Decoupling `muon_epsilon` lets the NS orthogonalization use a sensible eps (~1e-7) independent of Adam's much smaller `epsilon`.

**Backward compatibility**

Fully backward compatible: `use_hyperball` defaults to `False` (sharding reads it via `getattr(..., False)`), and `muon_epsilon` defaults to `adam_epsilon`. Existing configs produce identical behavior.

### 是否引起精度变化

否

devPR:https://github.com/PaddlePaddle/Paddle/pull/79792

### CODE REVIEW

**Comment by risemeup1111** [comment]
### Paddle-Bot Review Board （review完成)

| 序号 | 位置 | 优先级 | 规则来源 | 状态 |
| --- | --- | --- | --- | --- |
| 1 | <a href="https://github.com/PaddlePaddle/Paddle/pull/79793#discussion_r4057092272">Muon 更新步长首次应用 lr&#95;ratio 的行为变化（作者说明有意对齐 develop，已接受）</a> | P2 | <a href="https://console.cloud.baidu-int.com/devops/icode/repos/baidu/paddle/efficiency-tools/blob/master/nyanpasu/packages/nyanpasu-github-reviewer/src/nyanpasu_github_reviewer/rules/review.md">默认规则</a> | 🟡 |


<sub>Powered by Nyanpasu claude with Opus 4&#46;8 默认推理级别&#44; please check the suggestions carefully&#46;</sub>

File: python/paddle/optimizer/muon.py [hunk 1]
```diff
@@ -856,11 +978,11 @@ def ortho_fn(m):
...
                 if with_decay and weight_decay > 0:
                     if find_master:
-                        master_weight.scale_(1.0 - lr * weight_decay)
+                        master_weight.scale_(1.0 - effective_lr * weight_decay)
                     else:
-                        param.scale_(1.0 - lr * weight_decay)
+                        param.scale_(1.0 - effective_lr * weight_decay)
 
-                final_step = orthogonal_update * lr
+                final_step = orthogonal_update * effective_lr
```
Comment (risemeup1111) [inline python/paddle/optimizer/muon.py:988]: P2

此处将 `final_step` 及第 981/983 行 weight decay 中的 `lr` 改为 `effective_lr = lr * lr_ratio`，使普通 Muon（非 Hyperball）2D 参数**首次**应用了 `lr_ratio`——改动前 `_muon_update_group` 完全未使用 `lr_ratio`（仅 `_adamw_update` 用到）。

- `lr_ratio` 为默认 `None` 时 `effective_lr == lr`，行为不变；
- 但当用户配置了 `lr_ratio` 时，所有 Muon 参数的更新步长（含 weight decay 缩放）都会随之改变。

这与 PR 描述中「完全向后兼容 / 是否引起精度变化：否」在 `lr_ratio != None` 场景下不符。请确认该改动为有意对齐 `_adamw_update` 的行为，并在描述中注明此场景存在数值变化，以免升级 release/3.4 的既有 `lr_ratio` 配置时产生非预期的精度差异。

**Review by risemeup1111 — COMMENTED** [review]
本次 `synchronize`（`12d8e17` → `0ae2d0a`）仅将 `_hyperball_apply._fro` 由 `sqrt(sum(x*x))` 改为 `paddle.norm(p='fro')`，数值等价且新增测试已覆盖 2D/3D、fp32/fp64 等价性，无新增问题。

此前提出的 P2（普通 Muon 2D 更新路径首次应用 `lr_ratio`，在 `lr_ratio != None` 时改变数值，与「无精度变化」描述不符）对应代码本次未改动、也未见回复，仍待确认，详见既有评论线程。

File: python/paddle/optimizer/muon.py (hunk 1 above)
Comment (xxyux) [reply python/paddle/optimizer/muon.py:988]: 该改动是有意的，目的有二：
  - 对齐 _adamw_update：_adamw_update 本就使用 lr_ratio，而 _muon_update_group 此前完全没用，两条路径对 lr_ratio 的处理不一致。本次把 Muon 2D 参数的 final_step 与 weight decay 缩放也改为 effective_lr = lr * lr_ratio，使两者一致。
  - 对齐 develop：develop 上 _muon_update_group 的这几行早已是 effective_lr（由更早一个独立提交引入）。release/3.4 的 base 因为那次重构从未 cherry-pick 过来，才仍停在裸 lr。本 PP 把该改动随 Hyperball 一并带到 release/3.4，是让 release/3.4 与 develop 收敛，而非新引入分叉。
  - 此外，Hyperball 的 per-group / 角学习率（经 lr_ratio 落地）也依赖 Muon 路径使用 effective_lr，否则 MuonH 的 lr_ratio 缩放不生效。

Comment (risemeup1111) [reply python/paddle/optimizer/muon.py:988]: 感谢说明，理由充分：与 `_adamw_update` 的 `lr_ratio` 处理对齐、向 develop 收敛（该几行在 develop 上早已是 `effective_lr`），且 MuonH 的 per-group/角学习率依赖此路径生效；默认 `lr_ratio=None` 时行为不变。此项按已接受处理，仅建议在 PR 描述中补一句该场景（`lr_ratio != None`）与 develop 收敛的说明，以便后续追溯。

**Review by risemeup1111 — APPROVED** [review]

**Review by GuoxiaWang — APPROVED** [review]

**Review by wanghuancoder — APPROVED** [review]
LGTM

**Review by swgu98 — APPROVED** [review]

**Review by sneaxiy — APPROVED** [review]
